### PR TITLE
Add Helm operator

### PR DESCRIPTION
**Pull request by cniackz**
### Objective:

To add `Helm Operator` for DirectPV

### Source:

* https://github.com/cniackz/directpv-helm-operator

### Testing Notes:

* https://github.com/cniackz/public/wiki/How-to-test-DirectPV-Operator-%E2%80%90-2

### Needed:

* https://github.com/minio/directpv/pull/821

### CODE REVIEW

**Review by balamurugana — COMMENTED** [review]
You could use helm from https://github.com/minio/directpv/tree/master/helm

**Comment by cniackz** [comment]
Sure @balamurugana I will be using `helm` from https://github.com/minio/directpv/tree/master/helm but also I need to test a little bit more prior the review. I will let you know when I am ready for the review.

**Comment by cniackz** [comment]
Two things:

1. Using the helm chart added previously, just moved from location.
2. Using latest tag 4.0.7 for the current Operator
3. Tested truly in k8s to deploy operator and install directpv

I think this is ready for another round of reviews.

File: helm/Makefile [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+BUNDLE_DEFAULT_CHANNEL := --default-channel=$(DEFAULT_CHANNEL)
+endif
+BUNDLE_METADATA_OPTS ?= $(BUNDLE_CHANNELS) $(BUNDLE_DEFAULT_CHANNEL)
+
+# IMAGE_TAG_BASE defines the docker.io namespace and part of the image name for remote images.
+# This variable is used to construct full image tags for bundle and catalog images.
+#
+# For example, running 'make bundle-build bundle-push catalog-build catalog-push' will build and push both
+# quay.io/directpv-operator-bundle:$VERSION and quay.io/directpv-operator-catalog:$VERSION.
+IMAGE_TAG_BASE ?= quay.io/cniackz4/directpv-operator
```
Comment (balamurugana) [inline helm/Makefile:32]: We would need to use `quay.io/minio/directpv-operator`

File: helm/Makefile
Comment (balamurugana) [inline helm/Makefile:1]: Can we have a bash script to generate operator bundle scripts and patch them with appropriate values for VERSION, IMAGE_TAG_BASE etc?

File: helm/PROJECT [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  manifests.sdk.operatorframework.io/v2: {}
+  scorecard.sdk.operatorframework.io/v2: {}
+projectName: directpv-operator
+resources:
+- api:
+    crdVersion: v1
+    namespaced: true
+  domain: quay.io
+  group: charts
+  kind: DirectpvChart
```
Comment (balamurugana) [inline helm/PROJECT:18]: Could we style this all lowercase `directpvchart` or pretty name `DirectPVChart`?

File: helm/README.md
Comment (balamurugana) [inline helm/README.md:1]: We could remove this file or have a minimal documentation.

File: helm/my-directpv.yaml
Comment (balamurugana) [inline helm/my-directpv.yaml:1]: can we move this to `examples` directory or put it into documentation?

File: helm/pvc.yaml
Comment (balamurugana) [inline helm/pvc.yaml:1]: can we move this to `examples` directory or put it into documentation?

File: helm/Makefile (hunk 1 above)
Comment (cniackz) [reply helm/Makefile:32]: I don't have access to push this image here. Who can help me? I will ask Harsha

File: helm/Makefile
Comment (cniackz) [reply helm/Makefile:1]: Hello @balamurugana, I have added the script `helm/generate.sh` 
Please check!.

**Comment by cniackz** [comment]
Waiting for tests to pass, then I will ask another round of reviews.

File: helm/generate.sh [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+}
+
+function main() {
+
+    # docker-build: Build docker image with the manager.
+    docker buildx build --platform linux/amd64 -t "${IMG}" .
+
+    # docker-push: Push docker image with the manager.
+    docker push "${IMG}"
```
Comment (balamurugana) [inline helm/generate.sh:51]: This command is required for release privilege. We can only keep other steps here.

File: helm/generate.sh [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    operator-sdk generate kustomize manifests -q
+    (cd /Users/cniackz/directpv/helm/config/manager && kustomize edit set image controller="$IMG")
+    kustomize build config/manifests | operator-sdk generate bundle "$BUNDLE_GEN_FLAGS"
+    operator-sdk bundle validate ./bundle
+
+    # bundle-build: Build the bundle image.
+    docker build -f bundle.Dockerfile -t "$BUNDLE_IMG" .
+
+    # bundle-push: Push the bundle image.
+    docker push "${BUNDLE_IMG}"
```
Comment (balamurugana) [inline helm/generate.sh:63]: This command is required for release privilege. We can only keep other steps here.

File: helm/generate.sh [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+    # docker-build: Build docker image with the manager.
+    docker buildx build --platform linux/amd64 -t "${IMG}" .
+
+    # docker-push: Push docker image with the manager.
+    docker push "${IMG}"
+
+    # bundle: Generate bundle manifests and metadata, then validate generated files.
+    operator-sdk generate kustomize manifests -q
+    (cd /Users/cniackz/directpv/helm/config/manager && kustomize edit set image controller="$IMG")
```
Comment (balamurugana) [inline helm/generate.sh:55]: Can we use relative path as `./helm/config/manager` ?

File: helm/generate.sh (hunk 3 above)
Comment (cniackz) [reply helm/generate.sh:51]: Hello @balamurugana what do you mean by `keep other steps here`, do you want me to remove this line from the script?, what is the release privilege?, what do you mean by that?. I just want to understand better your comment to proceed accordingly.

Comment (cniackz) [reply helm/generate.sh:51]: Sorry I got it now, will remove it then 👍

File: helm/generate.sh (hunk 5 above)
Comment (cniackz) [reply helm/generate.sh:56]: Oh my bad, I was testing this in my laptop and totally forgot to use relative path. I am sorry about that, I will fix it!.

**Comment by cniackz** [comment]
Waiting for tests to pass, then I will ask another round of reviews.

**Review by balamurugana — DISMISSED** [review]

File: helm/examples/pvc.yaml [hunk 6]
```diff
@@ -0,0 +1,16 @@
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  annotations:
+    volume.beta.kubernetes.io/storage-provisioner: directpv-min-io
+    volume.kubernetes.io/selected-node: minio-k8s7
+  name: please-remove-me-cesar
```
Comment (Praveenrajmani) [inline helm/examples/pvc.yaml:7]: Do we need this example? If so, can we use https://github.com/minio/directpv/blob/master/functests/sleep.yaml instead?

File: helm/helm.sh
[the remark this answers is not in the document]
Comment (cniackz) [reply helm/helm.sh:1]: Thank you @Praveenrajmani for your review!

**Comment by cniackz** [comment]
Waiting for tests to pass, then I ask for another review.

**Comment by cniackz** [comment]
Failures are not related to this PR, look: https://github.com/minio/directpv/pull/828

**Review by balamurugana — APPROVED** [review]

**Review by Praveenrajmani — APPROVED** [review]

**Pull request merged**